standard_rules/unusual_login.py

```diff
@@ -19,62 +23,80 @@ def rule(event):
+    GEO_INFO = geoinfo_from_ip(event.udm("source_ip"))
+
+    # As of Panther 1.19, mocking returns all mocked objects in a string
+    # GEO_INFO must be converted back to a dict to mimic the API call
+    if isinstance(GEO_INFO, str):
+        GEO_INFO = json.loads(GEO_INFO)
+
+    # Look up history of unique geolocations
     event_key = get_key(event)
-    last_login_info = get_string_set(event_key)
-    fingerprint_timestamp = str(datetime.datetime.now())
-    if not last_login_info:
-        # Store this as the first login if we've never seen this user login before
-        put_string_set(event_key, [json.dumps({login_tuple: fingerprint_timestamp})])
+    # Mocked during unit testing
+    previous_geo_logins = get_string_set(event_key)
+
+    # As of Panther 1.19, mocking returns all mocked objects in a string
+    # previous_geo_logins must be converted back to a set to mimic the API call
+    if isinstance(previous_geo_logins, str):
+        print("previous_geo_logins is a mocked string:")
+        print(previous_geo_logins)
+        previous_geo_logins = ast.literal_eval(previous_geo_logins)
+        print("new type of previous_geo_logins is:", type(previous_geo_logins))
+
+    new_login_geo = (
+        f"{GEO_INFO.get('region', '<UNKNOWN_REGION>')}"
+        ":"
+        f"{GEO_INFO.get('city', '<UNKNOWN_CITY>')}"
+    )
+    new_login_timestamp = event.get("p_event_time", "")
+
+    # convert set of single string to dictionary
+    previous_geo_logins = json.loads(previous_geo_logins.pop())
+
+    # don't alert if the geo is already in the history
+    if previous_geo_logins.get(new_login_geo):
+        # update timestamp of the existing geo in the history
+        previous_geo_logins[new_login_geo] = new_login_timestamp
+        # exclude Dynamo API call from unit test
+        if "mock" not in event:
+            # write the dictionary of geolocs:timestamps back to Dynamo
+            put_string_set(event_key, [json.dumps(previous_geo_logins)])
         return False
-    last_login_info = json.loads(last_login_info.pop())
 
-    # update the timestamp associated with this fingerprint
-    last_login_info[login_tuple] = fingerprint_timestamp
-    put_string_set(event_key, [json.dumps(last_login_info)])
+    # fire an alert when there are more unique geolocs:timestamps in the login history
+    # add a new geo to the dictionary
+    updated_geo_logins = previous_geo_logins
+    updated_geo_logins[new_login_geo] = new_login_timestamp
+
+    # remove the oldest geo from the history if the updated dict exceeds the
+    # specified history length
+    if len(updated_geo_logins) > GEO_HISTORY_LENGTH:
+        oldest = updated_geo_logins[new_login_geo]
+        for geo, time in updated_geo_logins.items():
+            if time < oldest:
+                oldest = time
+                oldest_login = geo
+        updated_geo_logins.pop(oldest_login)
+        print("updated_geo_logins:")
+        print(updated_geo_logins)
```

Comment:
Before merging lets make sure all these print statements are either removed or switched to debug statements.